Toolbox/arcsdm/pca.py

```diff
@@ -0,0 +1,186 @@
+""" ArcSDM 6 ToolBox for ArcGIS Pro
+
+Conversion and tool development for ArcGIS Pro by Geological Survey of Finland (GTK), 2024.
+
+Compute defined number of principal components for numeric input data and transform the data.
+
+Before computation, data is scaled according to specified scaler and NaN values removed or replaced.
+Optionally, a nodata value can be given to handle similarly as NaN values.
+
+This tool is based on the PCA implementation in the scikit-learn library originally developed by University of Turku.
+"""
+
+import arcpy
+import numpy as np
+import pandas as pd
+from sklearn.decomposition import PCA
+from sklearn.preprocessing import MinMaxScaler, RobustScaler, StandardScaler
+
+SCALERS = {"standard": StandardScaler, "min_max": MinMaxScaler, "robust": RobustScaler}
+
+def Execute(self, parameters, messages):
+    """The source code of the tool."""
+    input_data = parameters[0].valueAsText.split(';')
+    number_of_components = parameters[1].value
+    scaler_type = parameters[2].valueAsText
+    nodata_handling = parameters[3].valueAsText
+    nodata_value = parameters[4].value
+    transformed_data_output = parameters[5].valueAsText
+    principal_components_output = parameters[6].valueAsText
+    explained_variances_output = parameters[7].valueAsText
+    explained_variance_ratios_output = parameters[8].valueAsText
+
+    datasets = []
+    for data in input_data:
+        # Load input data
+        if arcpy.Describe(data).dataType == "RasterDataset" or arcpy.Describe(data).dataType == "RasterLayer":
+            data = arcpy.RasterToNumPyArray(data)
+        elif arcpy.Describe(input_data).dataType == "Table":
+            data = pd.DataFrame(arcpy.da.TableToNumPyArray(data, "*"))
+        elif arcpy.Describe(data).dataType == "FeatureLayer":
+            data = arcpy.da.FeatureClassToNumPyArray(data, "*")
+        datasets.append(data)
+    
+    stacked_arrays = np.stack(datasets, axis=0)
+
+    # Perform PCA
+    transformed_data, principal_components, explained_variances, explained_variance_ratios = compute_pca(
+        stacked_arrays, number_of_components, scaler_type, nodata_handling, nodata_value
+    )
+
+    arcpy.AddMessage('='*5 + ' PCA results ' + '='*5)
+    # Save output data
+    if isinstance(transformed_data, np.ndarray):
+        if transformed_data.ndim is 2 or transformed_data.ndim is 3:
+            arcpy.NumPyArrayToRaster(transformed_data).save(transformed_data_output)
+        else:
+            arcpy.da.NumPyArrayToTable(transformed_data, transformed_data_output)
+        arcpy.AddMessage(f'Transformed data is saved in {transformed_data_output}')
+
+        arcpy.AddMessage(f'Principal components {principal_components}')
+            
+        if explained_variances.ndim is 2 or explained_variances.ndim is 3:
+            arcpy.NumPyArrayToRaster(explained_variances).save(explained_variances_output)
```

Comment:
I don't think these other outputs need the dimension checks and saving to raster, since they shouldn't ever be anything except 1-dimensional arrays. In the eis_toolkit client functions, they're saved to a json file.